lib/rust/profiler/src/lib.rs

```diff
@@ -0,0 +1,333 @@
+        pub fn new() -> Self {
+            Self(cell::UnsafeCell::new(vec![]))
+        }
+
+        /// Push an element.
+        pub fn push(&self, element: T) {
+            // Note [LocalVecBuilder Safety]
+            unsafe {
+                (&mut *self.0.get()).push(element);
+            }
+        }
+
+        // Note [LocalVecBuilder Safety]
+        // =============================
+        // When obtaining a reference from the UnsafeCell, all accessors follow these rules:
+        // - There must be a scope that the reference doesn't escape.
+        // - There must be no other references obtained in the same scope.
+        // Consistently following these rules ensures the no-alias rule of mutable references is
+        // satisfied.
+
+        /// Return (and consume) all elements pushed so far.
+        pub fn build(&self) -> Vec<T> {
+            // Note [LocalVecBuilder Safety]
+            unsafe { mem::take(&mut *self.0.get()) }
+        }
+    }
+
+    // ====================
+    // === MEASUREMENTS ===
+    // ====================
+
+    thread_local! {
+        pub static MEASUREMENTS: LocalVecBuilder<Measurement> = LocalVecBuilder::new();
+    }
+}
+
+/// Global log of Measurements.
+pub use internal::MEASUREMENTS;
+
+/// Gather all measurements taken since the last time take_log() was called.
+pub fn take_log() -> Vec<Measurement> {
+    MEASUREMENTS.with(|log| log.build())
+}
+
+
+// ================
+// === Profiler ===
+// ================
+
+/// The interface supported by profilers of all profiling levels.
+pub trait Profiler: Copy {
+    type Started: StartedProfiler;
+    fn finish(self, data: &Self::Started);
+}
+
+
+
+// ==============
+// === Parent ===
+// ==============
+
+/// Any object representing a profiler that is a valid parent for a profiler of type T.
+pub trait Parent<T: Profiler> {
+    fn new_child(&self, label: Label) -> Started<T>;
+    fn new_child_same_start(&self, label: Label) -> Started<T>;
+}
+
+
+
+// ===============
+// === Started ===
+// ===============
+
+/// A profiler that has a start time set, and will complete its measurement when dropped.
+pub struct Started<T: Profiler> {
+    profiler: T,
+    data:     T::Started,
+}
+
+
+// === Trait Implementations ===
+
+impl<T: Profiler> Drop for Started<T> {
+    fn drop(&mut self) {
+        let profiler = self.profiler;
+        profiler.finish(&self.data);
+    }
+}
+
+impl<T, U> Parent<T> for Started<U>
+where
+    U: Parent<T> + Profiler,
+    T: Profiler,
+{
+    fn new_child(&self, label: Label) -> Started<T> {
+        self.profiler.new_child(label)
+    }
+
+    fn new_child_same_start(&self, label: Label) -> Started<T> {
+        self.profiler.new_child_same_start(label)
+    }
+}
+
+
+
+// ===================================
+// === profiler_macros Invocations ===
+// ===================================
+
+#[doc(inline)]
+pub use profiler_macros::profile;
+
+profiler_macros::define_hierarchy![Objective, Task, Detail, Debug];
```

Comment:
This macro is not proc-macro. Is the expansion understood by IntellIJ / CLion – are the created structures hinted to the user? For example, if I write `profiler::` am I hinted with different possible profiler hierarchy definitions? It is important to hint the user of this API as much as possible. If these are not hinted to the user, I would vote for moving the generation of the new structures to proc-macros and leaving only the `#[profile]` macro as a non-proc-macro definition.

Also, as we've been speaking during designing of this task, in case the user is not hinted with anything when using IntelliJ, we should just use proc-macros even if they are uglier. It applies to the `#[profile]` vs `profile!(...)` syntax as well. If the `#[profile]` syntax prevents the user of this API from any hints, we should use proc-macros instead.

Reply:
1. This *is* a proc macro--I think you mean it is not a declarative macro. Declarative macros are the simpler `macro_rules!` macros; proc macros are the kind that explicitly operate on a `TokenStream`, that JetBrains only recently supports.
2. It is [working for me](https://i.imgur.com/cTUKSd3.png), in CLion 2021.3.1 (released Dec 28). Are you using the latest release of IntelliJ?

Reply:
1. Ugh, I'm always mixing the names here. OF course, I meant `macro-rules` here instead.
2. I was not testing it TBH. I know it was not working a few months ago, so I preferred to ask. If it works now, then great! (I hope you did not read my message as "it does not work, lets fix it!" – it was a question about IF it works with explanation why its important to me) :)

Reply:
So, **`define_hierarchy!` is fully working** for me: CLion seems to know the expansion of that proc macro; it suggests the generated types, has documentation for them, etc.

However I just noticed a problem with `#[profile]`, and using the `profile!` syntax doesn't help: *within* the decorated function, hints are working as usual; but the IDE doesn't seem to know about the *output*, i.e. **adding profiling to a function causes the IDE no longer to know about the function's definition**. This is [supposed to be working](https://github.com/intellij-rust/intellij-rust/issues/6908), but it looks like there are still gaps in their implementation.

# What does JetBrains support?

Doing some testing, it looks like the IDE understands *most* items generated by a proc macro, including everything `define_hierarchy!` needs:
- structs
- macros
- functions defined in impl blocks

What isn't working is proc macros that expand to *a single function*, rather than an item that can contain functions.

# Alternative APIs that avoid the bug

## Use `profile!(fn ..)` syntax; implement with a declarative macro

Pros:
- It would work with current tooling. (The first step would be to do a mock implementation to verify this).

Cons:
- The syntax is not as nice.

Complexity: I think a declarative implementation that supports everything we need to support would be tricky. I could probably implement it in 3-4 days.

## Apply `#[profile]` to the outer objects

JetBrains seems to be okay with macros generating impl blocks, so we could apply `#[profile]` at a higher scope than individual `fn`s.

Example:
```
struct Foo;

#[profile]
impl Foo {
    fn profiled_fn1(&self, [profiler: profiler::Task]) { .. }
    fn profiled_fn2(x: u32, [profiler: profiler::Objective]) { .. }
    fn not_profiled(self, x: u32) { .. }
}
```

(In this syntax, profiler arguments are marked with brackets, to make it more obvious where the macro is changing signatures).

Pros:
- It would work with current tooling. (I have already tested this: the IDE knows about functions emitted inside struct impls, trait impls, and modules).
- Unobtrusive syntax
- When JetBrains fixes the bug, we could also enable annotating at `fn` granularity

Cons:
- Moving the attribute farther from the functions it affects makes it less obvious what's going on with the `profiler` arguments.

Complexity: This would be simpler than the declarative implementation; I think I could implement and document it 2 days.

# Proceeding

For `#[profile]`, we have design change and implementation work needed. The lower-level `define_hierarchy!` is stable; I don't anticipate needing any changes to accommodate whatever we do with `#[profile]`. Should I cut the `#[profile]` code out of this PR so we can get the low-level code merged, and tackle the high-level interface separately? (If we go with the declarative `profile!` API, that will require a rewritten implementation anyway)